### ISSUE TITLE
command ```zfs list -t filesystem,snapshot <dataset>``` not work as it described in documentation

### ISSUE THREAD

**Issue by makhomed** [openzfs/zfs#13657]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Rocky Linux
Distribution Version	| release 8.6 (Green Obsidian)
Kernel Version	| 4.18.0-372.13.1.el8_6.x86_64
Architecture	| x86_64
OpenZFS Version	| zfs-2.1.5-1

### Describe the problem you're observing

Command ```zfs list``` not work as it described in documentation:
```
     -t type
         A comma-separated list of types to display, where type is one of filesystem, snapshot, volume, bookmark, or all.  
         For example, specifying -t snapshot displays only snapshots.
```
Command ```zfs list -t filesystem,snapshot <dataset>``` or ```zfs list -t all <dataset>``` don't display snapshots:
```
# zfs list -t filesystem,snapshot tank/mirror/core9/containers
NAME                           USED  AVAIL     REFER  MOUNTPOINT
tank/mirror/core9/containers  42.4G  63.2T      227K  /tank/mirror/core9/containers

# zfs list -t all tank/mirror/core9/containers
NAME                           USED  AVAIL     REFER  MOUNTPOINT
tank/mirror/core9/containers  42.4G  63.2T      227K  /tank/mirror/core9/containers
```
Workaround ```zfs list -d1 -r -t filesystem,snapshot <dataset>``` is not possible, because this command display also child datasets - this is not expected for command ```zfs list -t all <dataset>``` - only dataset itself and it snapshots expected in the list, not child datasets.

So, this is not bug in zfs documentation, this is bug in zfs code, and no known workarounds exists for this bug.

### Describe how to reproduce the problem

Create any filesystem with snapshots and child filesystems, and run commands
```zfs list -t filesystem,snapshot <dataset>``` or ```zfs list -t all <dataset>``` on it.

### Include any warning/errors/backtraces from the system logs

No any warning/errors/backtraces in the system logs.

**Comment by makhomed** [comment]
> `zfs list -t snapshot -r -d1 [filesystem|volume]` appears to do what you want.

No, because command `zfs list -t snapshot -r -d1 <dataset>` also display child datasets.
This is unwanted and unexpected behaviour and this command can't be accepted as workaround for the bug.

Commands `zfs list -t filesystem <dataset>` and `zfs list -t snapshot <dataset>` don't display child datasets, as it expected.

Command `zfs list -t filesystem,snapshot <dataset>` expected to work as logical combination of two abobe commands.

But command `zfs list -t filesystem,snapshot <dataset>` don't display snapshots as it described in documentation `man zfs-list`:

```
-t type
A comma-separated list of types to display, where type is one of filesystem, snapshot, volume, bookmark, or all.  
For example, specifying -t snapshot displays only snapshots.
```

> I don't see how the description is inconsistent

Description of `-t` switch at `man zfs-list` is ok, there is no bugs in documentation, bugs are in ZFS code:
Commands `zfs list -t all <dataset>` and `zfs list -t filesystem,snapshot <dataset>` don't display snapshots. 

> snapshots are treated as children, hierarchically, of the dataset they're of, so if you ask non-recursively for all dataset types at `foo/bar`, it's going to give you `foo/bar`, not `foo/bar/baz` or `foo/bar@bat` or `foo/bar#bing`. (I'm not saying you're right or wrong, just that I don't understand why you say this is a bug.)

Command  `zfs list -t snapshot <dataset>`  work exactly as it described in documentation - there is no bugs here.
But command `zfs list -t all <dataset>` and `zfs list -t filesystem,snapshot <dataset>` don't displat snapshots. This is bug.
 
> (I think this is inconsistent with the special casing of `zfs list -t snapshot [filesystem|volume]` to be equivalent to the first command I mentioned, but I think we're stuck with that at this point even if we wanted to change it, because people expect it now...)

Command  `zfs list -t snapshot <dataset>`  work as expected and as it described in documentation, all ok with this case.

> snapshots are treated as children

In zfs internals - yes, but in zfs command line no: `zfs list -t snapshot <dataset>` work ok without extra `-r -d1` switches.

What is the problem to fix commands `zfs list -t all <dataset>` and `zfs list -t filesystem,snapshot <dataset>` 
to work as it described in the zfs documentation and analogically to `zfs list -t snapshot <dataset>` command?

**Comment by rincebrain** [comment]
First, I should point out that unless your system has a bug mine does not, my initial command does not display non-snapshot children. I tested it before telling you, and while your command with `-t filesystem,snapshot` does do that, mine does not.

It seems like what you're objecting to is:
* `zfs list -t snapshot [dataset]` cheats and rewrites itself internally to `zfs list -t snapshot -r -d1 [dataset]`
* You would like `zfs list -t filesystem,snapshot` to be the combination of `zfs list -t filesystem -d0 [filesystem]` and `zfs list -t snapshot [filesystem]`, the latter of which acts like you had `-r -d1` on the command line as well

I don't think extending the special casing is a good idea, because that would break people's expectations of how the rest of the system works.

I also don't think this is a bug in the code, just an unfortunate inconsistent choice that was made and is not documented well enough.

But perhaps other people will disagree with me, I'm not the final arbiter of anything.

**Comment by rincebrain** [comment]
I didn't suggest it to cause problems, my frustrated remark about the situation aside.

If people's expectations and the behavior are inconsistent, that's probably a bug if the people's expectations are consistent and reasonable.
If the tool is inconsistent in different usages, then that's a bug, as well.

This disconnect is because someone added behavior that is inconsistent with how the tool always worked in one special case, though it is convenient.

Changing the behavior to do that more globally seems like it would cause a great deal of strange fallout and breakage, much moreso than removing the special case.
Adding documentation can't resolve this on its own because people often don't read it, or don't understand it, and you can't document your way out of "the tool does something completely different if you change something apparently unrelated", much like it would still be unreasonable if running "checkers --colors=white,black" launched a game of solitaire.

So removing the special case seems like the least disruptive solution, to me. But I'm not authoritative or able to make any decisions, I'm just remarking on what I think the least painful solution that would avoid maximum difficulties would be.

I don't object to updating the documentation to try and clarify that this expectation is inconsistent from everything else, but I don't think there's any amount of documentation that will avoid people rightly pointing out that this is very inconsistent.

**Comment by rincebrain** [comment]
I'm sorry you think further conversation is unconstructive - nothing I said is written in stone, it's just my current perspective, which doesn't matter that much, in the grand scheme of things, in terms of how it influences what people with actual authority choose to do on the project.

I was just trying to explain why it seemed like the least disruptive solution to me, even though it's also the most inconvenient for me, as I too use that shortcut regularly. But as I said above, I think just adding more documentation is insufficient for how inconsistent this behavior is, so either way you're going to be breaking someone's expectations now.

**Comment by makhomed** [comment]
> ...so you're breaking someone's expectations anyway.

I think what further conversation is unconstructive because it is no good solution for this case, acceptable for all.

People with actual authority choose to do on the project already marked this situation in https://github.com/openzfs/zfs/issues/10130 as documentation defect.

So, this issue should be closed as duplicate.

Thank you for your time.
